// contentful-management/dist/es-modules/entities/app-installation.js
import { toPlainObject, freezeSys } from 'contentful-sdk-core';
import copy from 'fast-copy';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createAppInstallationApi(http) {
  var getParams = function getParams(data) {
    return {
      spaceId: data.sys.space.sys.id,
      environmentId: data.sys.environment.sys.id,
      appDefinitionId: data.sys.appDefinition.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.appInstallation.upsert(http, getParams(data), data).then(function (data) {
        return wrapAppInstallation(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.appInstallation.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw App Installation data
 * @return Wrapped App installation data
 */


export function wrapAppInstallation(http, data) {
  var appInstallation = toPlainObject(copy(data));
  var appInstallationWithMethods = enhanceWithMethods(appInstallation, createAppInstallationApi(http));
  return freezeSys(appInstallationWithMethods);
}
/**
 * @private
 */

export var wrapAppInstallationCollection = wrapCollection(wrapAppInstallation);
