// contentful-management/dist/es-modules/entities/app-definition.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createAppDefinitionApi(http) {
  var getParams = function getParams(data) {
    return {
      appDefinitionId: data.sys.id,
      organizationId: data.sys.organization.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.appDefinition.update(http, getParams(data), data).then(function (data) {
        return wrapAppDefinition(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.appDefinition.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw App Definition data
 * @return Wrapped App Definition data
 */


export function wrapAppDefinition(http, data) {
  var appDefinition = toPlainObject(copy(data));
  var appDefinitionWithMethods = enhanceWithMethods(appDefinition, createAppDefinitionApi(http));
  return freezeSys(appDefinitionWithMethods);
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw App Definition collection data
 * @return Wrapped App Definition collection data
 */

export var wrapAppDefinitionCollection = wrapCollection(wrapAppDefinition);
