// contentful-management/dist/es-modules/entities/editor-interface.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createEditorInterfaceApi(http) {
  return {
    update: function update() {
      var self = this;
      var raw = self.toPlainObject();
      return endpoints.editorInterface.update(http, {
        spaceId: self.sys.space.sys.id,
        environmentId: self.sys.environment.sys.id,
        contentTypeId: self.sys.contentType.sys.id
      }, raw).then(function (response) {
        return wrapEditorInterface(http, response);
      });
    },
    getControlForField: function getControlForField(fieldId) {
      var self = this;
      var result = (self.controls || []).filter(function (control) {
        return control.fieldId === fieldId;
      });
      return result && result.length > 0 ? result[0] : null;
    }
  };
}
/**
 * @private
 */


export function wrapEditorInterface(http, data) {
  var editorInterface = toPlainObject(copy(data));
  var editorInterfaceWithMethods = enhanceWithMethods(editorInterface, createEditorInterfaceApi(http));
  return freezeSys(editorInterfaceWithMethods);
}
/**
 * @private
 */

export var wrapEditorInterfaceCollection = wrapCollection(wrapEditorInterface);
