// contentful-management/dist/es-modules/entities/environment-alias.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createEnvironmentAliasApi(http) {
  var getParams = function getParams(alias) {
    return {
      spaceId: alias.sys.space.sys.id,
      environmentAliasId: alias.sys.id
    };
  };

  return {
    update: function update() {
      var raw = this.toPlainObject();
      return endpoints.environmentAlias.update(http, getParams(raw), raw).then(function (data) {
        return wrapEnvironmentAlias(http, data);
      });
    },
    "delete": function _delete() {
      var raw = this.toPlainObject();
      return endpoints.environmentAlias.del(http, getParams(raw)).then(function () {// noop
      });
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw environment alias data
 * @return Wrapped environment alias data
 */


export function wrapEnvironmentAlias(http, data) {
  var alias = toPlainObject(copy(data));
  var enhancedAlias = enhanceWithMethods(alias, createEnvironmentAliasApi(http));
  return freezeSys(enhancedAlias);
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw environment alias collection data
 * @return Wrapped environment alias collection data
 */

export var wrapEnvironmentAliasCollection = wrapCollection(wrapEnvironmentAlias);
