// contentful-management/dist/es-modules/entities/organization-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createOrganizationMembershipApi(http, organizationId) {
  var getParams = function getParams(data) {
    return {
      organizationMembershipId: data.sys.id,
      organizationId: organizationId
    };
  };

  return {
    update: function update() {
      var raw = this.toPlainObject();
      return endpoints.organizationMembership.update(http, getParams(raw), raw).then(function (data) {
        return wrapOrganizationMembership(http, data, organizationId);
      });
    },
    "delete": function del() {
      var raw = this.toPlainObject();
      return endpoints.organizationMembership.del(http, getParams(raw));
    }
  };
}
/**
 * @private
 * @param {Object} http - HTTP client instance
 * @param {Object} data - Raw organization membership data
 * @return {OrganizationMembership} Wrapped organization membership data
 */


export function wrapOrganizationMembership(http, data, organizationId) {
  var organizationMembership = toPlainObject(copy(data));
  var organizationMembershipWithMethods = enhanceWithMethods(organizationMembership, createOrganizationMembershipApi(http, organizationId));
  return freezeSys(organizationMembershipWithMethods);
}
/**
 * @private
 */

export var wrapOrganizationMembershipCollection = wrapCollection(wrapOrganizationMembership);
