// contentful-management/dist/es-modules/entities/role.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createRoleApi(http) {
  var getParams = function getParams(data) {
    return {
      spaceId: data.sys.space.sys.id,
      roleId: data.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.role.update(http, getParams(data), data).then(function (data) {
        return wrapRole(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.role.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw role data
 * @return Wrapped role data
 */


export function wrapRole(http, data) {
  var role = toPlainObject(copy(data));
  var roleWithMethods = enhanceWithMethods(role, createRoleApi(http));
  return freezeSys(roleWithMethods);
}
/**
 * @private
 */

export var wrapRoleCollection = wrapCollection(wrapRole);
