// contentful-management/dist/es-modules/entities/space-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createSpaceMembershipApi(http) {
  var getParams = function getParams(data) {
    return {
      spaceId: data.sys.space.sys.id,
      spaceMembershipId: data.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.spaceMembership.update(http, getParams(data), data).then(function (data) {
        return wrapSpaceMembership(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.spaceMembership.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw space membership data
 * @return Wrapped space membership data
 */


export function wrapSpaceMembership(http, data) {
  var spaceMembership = toPlainObject(copy(data));
  var spaceMembershipWithMethods = enhanceWithMethods(spaceMembership, createSpaceMembershipApi(http));
  return freezeSys(spaceMembershipWithMethods);
}
/**
 * @private
 */

export var wrapSpaceMembershipCollection = wrapCollection(wrapSpaceMembership);
