// contentful-management/dist/es-modules/entities/team.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

/**
 * @private
 */
function createTeamApi(http) {
  var getParams = function getParams(data) {
    return {
      teamId: data.sys.id,
      organizationId: data.sys.organization.sys.id
    };
  };

  return {
    update: function update() {
      var raw = this.toPlainObject();
      return endpoints.team.update(http, getParams(raw), raw).then(function (data) {
        return wrapTeam(http, data);
      });
    },
    "delete": function del() {
      var raw = this.toPlainObject();
      return endpoints.team.del(http, getParams(raw));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw team data
 * @return Wrapped team data
 */


export function wrapTeam(http, data) {
  var team = toPlainObject(copy(data));
  var teamWithMethods = enhanceWithMethods(team, createTeamApi(http));
  return freezeSys(teamWithMethods);
}
/**
 * @private
 */

export var wrapTeamCollection = wrapCollection(wrapTeam);
