// contentful-management/dist/es-modules/entities/team-space-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createTeamSpaceMembershipApi(http) {
  var getParams = function getParams(data) {
    return {
      teamSpaceMembershipId: data.sys.id,
      spaceId: data.sys.space.sys.id
    };
  };

  return {
    update: function update() {
      var raw = this.toPlainObject();
      return endpoints.teamSpaceMembership.update(http, getParams(raw), raw).then(function (data) {
        return wrapTeamSpaceMembership(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.teamSpaceMembership.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw space membership data
 * @return Wrapped team space membership data
 */


export function wrapTeamSpaceMembership(http, data) {
  var teamSpaceMembership = toPlainObject(copy(data));
  var teamSpaceMembershipWithMethods = enhanceWithMethods(teamSpaceMembership, createTeamSpaceMembershipApi(http));
  return freezeSys(teamSpaceMembershipWithMethods);
}
/**
 * @private
 */

export var wrapTeamSpaceMembershipCollection = wrapCollection(wrapTeamSpaceMembership);
