// contentful-management/dist/es-modules/entities/team-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
import * as endpoints from '../plain/endpoints';

function createTeamMembershipApi(http) {
  var getParams = function getParams(data) {
    return {
      teamMembershipId: data.sys.id,
      teamId: data.sys.team.sys.id,
      organizationId: data.sys.organization.sys.id
    };
  };

  return {
    update: function update() {
      var raw = this.toPlainObject();
      return endpoints.teamMembership.update(http, getParams(raw), raw).then(function (data) {
        return wrapTeamMembership(http, data);
      });
    },
    "delete": function del() {
      var raw = this.toPlainObject();
      return endpoints.teamMembership.del(http, getParams(raw));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw team membership data
 * @return Wrapped team membership data
 */


export function wrapTeamMembership(http, data) {
  var teamMembership = toPlainObject(copy(data));
  var teamMembershipWithMethods = enhanceWithMethods(teamMembership, createTeamMembershipApi(http));
  return freezeSys(teamMembershipWithMethods);
}
/**
 * @private
 */

export var wrapTeamMembershipCollection = wrapCollection(wrapTeamMembership);
