// contentful-management/dist/es-modules/entities/ui-extension.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import * as endpoints from '../plain/endpoints';
import { wrapCollection } from '../common-utils';

function createUiExtensionApi(http) {
  var getParams = function getParams(data) {
    return {
      spaceId: data.sys.space.sys.id,
      environmentId: data.sys.environment.sys.id,
      extensionId: data.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.uiExtension.update(http, getParams(data), data).then(function (data) {
        return wrapUiExtension(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.uiExtension.del(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw UI Extension data
 * @return Wrapped UI Extension data
 */


export function wrapUiExtension(http, data) {
  var uiExtension = toPlainObject(copy(data));
  var uiExtensionWithMethods = enhanceWithMethods(uiExtension, createUiExtensionApi(http));
  return freezeSys(uiExtensionWithMethods);
}
/**
 * @private
 */

export var wrapUiExtensionCollection = wrapCollection(wrapUiExtension);
