// contentful-management/dist/es-modules/entities/webhook.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import * as endpoints from '../plain/endpoints';
import { wrapCollection } from '../common-utils';

function createWebhookApi(http) {
  var getParams = function getParams(data) {
    return {
      spaceId: data.sys.space.sys.id,
      webhookDefinitionId: data.sys.id
    };
  };

  return {
    update: function update() {
      var data = this.toPlainObject();
      return endpoints.webhook.update(http, getParams(data), data).then(function (data) {
        return wrapWebhook(http, data);
      });
    },
    "delete": function del() {
      var data = this.toPlainObject();
      return endpoints.webhook.del(http, getParams(data));
    },
    getCalls: function getCalls() {
      var data = this.toPlainObject();
      return endpoints.webhook.getManyCallDetails(http, getParams(data));
    },
    getCall: function getCall(id) {
      var data = this.toPlainObject();
      return endpoints.webhook.getCallDetails(http, _objectSpread(_objectSpread({}, getParams(data)), {}, {
        callId: id
      }));
    },
    getHealth: function getHealth() {
      var data = this.toPlainObject();
      return endpoints.webhook.getHealthStatus(http, getParams(data));
    }
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw webhook data
 * @return Wrapped webhook data
 */


export function wrapWebhook(http, data) {
  var webhook = toPlainObject(copy(data));
  var webhookWithMethods = enhanceWithMethods(webhook, createWebhookApi(http));
  return freezeSys(webhookWithMethods);
}
/**
 * @private
 */

export var wrapWebhookCollection = wrapCollection(wrapWebhook);
