// contentful-management/dist/es-modules/plain/endpoints/content-type.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as raw from './raw';
import { normalizeSelect } from './utils';
import copy from 'fast-copy';

var getBaseUrl = function getBaseUrl(params) {
  return "/spaces/".concat(params.spaceId, "/environments/").concat(params.environmentId, "/content_types");
};

var getContentTypeUrl = function getContentTypeUrl(params) {
  return getBaseUrl(params) + "/".concat(params.contentTypeId);
};

export var get = function get(http, params) {
  return raw.get(http, getContentTypeUrl(params), {
    params: normalizeSelect(params.query)
  });
};
export var getMany = function getMany(http, params) {
  return raw.get(http, getBaseUrl(params), {
    params: params.query
  });
};
export var create = function create(http, params, rawData) {
  var data = copy(rawData);
  return raw.post(http, getBaseUrl(params), data);
};
export var createWithId = function createWithId(http, params, rawData) {
  var data = copy(rawData);
  return raw.put(http, getContentTypeUrl(params), data);
};
export var update = function update(http, params, rawData, headers) {
  var _rawData$sys$version;

  var data = copy(rawData);
  delete data.sys;
  return raw.put(http, getContentTypeUrl(params), data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export var del = function del(http, params) {
  return raw.del(http, getContentTypeUrl(params));
};
export var publish = function publish(http, params, rawData) {
  return raw.put(http, getContentTypeUrl(params) + '/published', null, {
    headers: {
      'X-Contentful-Version': rawData.sys.version
    }
  });
};
export var unpublish = function unpublish(http, params) {
  return raw.del(http, getContentTypeUrl(params) + '/published');
};

/**
 * @private
 * @param id - unique ID of the field
 * @param key - the attribute on the field to change
 * @param value - the value to set the attribute to
 */
var findAndUpdateField = function findAndUpdateField(contentType, fieldId, omitOrDelete) {
  var field = contentType.fields.find(function (field) {
    return field.id === fieldId;
  });

  if (!field) {
    return Promise.reject(new Error("Tried to omitAndDeleteField on a nonexistent field, ".concat(fieldId, ", on the content type ").concat(contentType.name, ".")));
  } // @ts-expect-error


  field[omitOrDelete] = true;
  return Promise.resolve(contentType);
};

export var omitAndDeleteField = function omitAndDeleteField(http, params, contentType, fieldId) {
  return findAndUpdateField(contentType, fieldId, 'omitted').then(function (newContentType) {
    return update(http, params, newContentType);
  }).then(function (newContentType) {
    return findAndUpdateField(newContentType, fieldId, 'deleted');
  }).then(function (newContentType) {
    return update(http, params, newContentType);
  });
};
