// contentful-management/dist/es-modules/plain/endpoints/index.js
import * as _contentType from './content-type';
export { _contentType as contentType };
import * as _entry from './entry';
export { _entry as entry };
import * as _asset from './asset';
export { _asset as asset };
import * as _environment from './environment';
export { _environment as environment };
import * as _environmentAlias from './environment-alias';
export { _environmentAlias as environmentAlias };
import * as _locale from './locale';
export { _locale as locale };
import * as _raw from './raw';
export { _raw as raw };
import * as _space from './space';
export { _space as space };
import * as _user from './user';
export { _user as user };
import * as _editorInterface from './editor-interface';
export { _editorInterface as editorInterface };
import * as _personalAccessToken from './personal-access-token';
export { _personalAccessToken as personalAccessToken };
import * as _usage from './usage';
export { _usage as usage };
import * as _role from './role';
export { _role as role };
import * as _scheduledAction from './scheduled-action';
export { _scheduledAction as scheduledAction };
import * as _previewApiKey from './preview-api-keys';
export { _previewApiKey as previewApiKey };
import * as _apiKey from './api-key';
export { _apiKey as apiKey };
import * as _upload from './upload';
export { _upload as upload };
import * as _appDefinition from './app-definition';
export { _appDefinition as appDefinition };
import * as _appInstallation from './app-installation';
export { _appInstallation as appInstallation };
import * as _uiExtension from './ui-extension';
export { _uiExtension as uiExtension };
import * as _webhook from './webhook';
export { _webhook as webhook };
import * as _tag from './tag';
export { _tag as tag };
import * as _organization from './organization';
export { _organization as organization };
import * as _organizationInvitation from './organization-invitation';
export { _organizationInvitation as organizationInvitation };
import * as _organizationMembership from './organization-membership';
export { _organizationMembership as organizationMembership };
import * as _snapshot from './snapshot';
export { _snapshot as snapshot };
import * as _team from './team';
export { _team as team };
import * as _teamMembership from './team-membership';
export { _teamMembership as teamMembership };
import * as _teamSpaceMembership from './team-space-membership';
export { _teamSpaceMembership as teamSpaceMembership };
import * as _spaceMember from './space-member';
export { _spaceMember as spaceMember };
import * as _spaceMembership from './space-membership';
export { _spaceMembership as spaceMembership };
